overlord/hookstate/ctlcmd/kmod.go

```diff
@@ -0,0 +1,201 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2022 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package ctlcmd
+
+import (
+	"fmt"
+
+	"github.com/snapcore/snapd/i18n"
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/osutil/kmod"
+	"github.com/snapcore/snapd/overlord/hookstate"
+	"github.com/snapcore/snapd/overlord/ifacestate"
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/snap"
+)
+
+var (
+	shortKmodHelp = i18n.G("Load or unload kernel modules")
+	longKmodHelp  = i18n.G(`
+The kmod command handles loading and unloading of kernel modules.`)
+)
+
+func init() {
+	addCommand("kmod", shortKmodHelp, longKmodHelp, func() command {
+		cmd := &kmodCommand{}
+		cmd.InsertCmd.kmodBaseCommand = cmd
+		cmd.RemoveCmd.kmodBaseCommand = cmd
+		return cmd
+	})
+}
+
+type kmodCommand struct {
+	baseCommand
+	InsertCmd KModInsertCmd `command:"insert" description:"load a kernel module"`
+	RemoveCmd KModRemoveCmd `command:"remove" description:"unload a kernel module"`
+}
+
+type kmodSubcommand struct {
+	kmodBaseCommand *kmodCommand
+	snapInfo        *snap.Info
+}
+
+type KModInsertCmd struct {
+	kmodSubcommand
+	Positional struct {
+		Module  string   `positional-arg-name:"<module>" required:"yes" description:"name of the kernel module to be loaded"`
+		Options []string `positional-arg-name:"<options>" description:"kernel module options"`
+	} `positional-args:"yes" required:"yes"`
+}
+
+func (k *KModInsertCmd) Execute([]string) error {
+	context, err := k.kmodBaseCommand.ensureContext()
```

Comment:
does this work?
```suggestion
	context, err := k.ensureContext()
```

OTOH does this need kmodBaseCommand/kmodSubCommand at all? the only 2 functions that are tied to kmodSubCommand could easily be generic helpers.

Reply:
This (your diff suggestion) cannot work, because the `KModInsertCmd` type encapsulates a `kmodSubCommand` type, but not a `kmodBaseCommand` (which is only available as a *pointer* inside `kmodSubCommand`). The fact that I have to store this pointer (see the `init()` function at line 40) is not very nice, and it may be that someone more expert with `go-flags` finds a better way to implement subcommands and passing data to them.

As for changing the two functions into helpers, that would indeed allow me to remove the `kmodSubCommand` type, but then I'd anyway need to replicate its contents in `KModInsertCmd` and `KModRemoveCmd`, and add some extra arguments to the functions. So I'm not sure it's a net win.

Reply:
Doesn't look too bad: https://paste.ubuntu.com/p/rKMKNqfZzw/